internal/controlplane/handlers_authz.go

```diff
@@ -355,16 +493,150 @@ func (s *Server) AssignRole(ctx context.Context, req *minder.AssignRoleRequest)
+	}
+	return nil, util.UserVisibleError(codes.InvalidArgument, "one of subject or email must be specified")
+}
 
-	if role == "" || sub == "" {
+func (s *Server) removeInvite(
+	ctx context.Context,
+	targetPrj uuid.UUID,
+	role authz.Role,
+	email string,
+) (*minder.RemoveRoleResponse, error) {
+	prj := targetPrj.String()
+	// Get all invitations for this email, project and role
+	inviteToRemove, err := s.store.GetInvitationByEmailAndProjectAndRole(ctx, db.GetInvitationByEmailAndProjectAndRoleParams{
+		Email:   email,
+		Project: targetPrj,
+		Role:    role.String(),
+	})
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, util.UserVisibleError(codes.NotFound, "no invitation found for this email, project and role")
+		}
+		return nil, status.Errorf(codes.Internal, "error getting invitation: %v", err)
+	}
+
+	// Delete the invitation
+	_, err = s.store.DeleteInvitation(ctx, inviteToRemove.Code)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error deleting invitation: %v", err)
+	}
+
+	// Return the response
+	return &minder.RemoveRoleResponse{
+		RoleAssignment: &minder.RoleAssignment{
+			Role:    role.String(),
+			Email:   email,
+			Project: &prj,
+		},
+	}, nil
+}
+
+func (s *Server) removeRole(
+	ctx context.Context,
+	targetProject uuid.UUID,
+	role authz.Role,
+	subject string,
+) (*minder.RemoveRoleResponse, error) {
+	var err error
+	// We may be given a human-readable identifier which can vary over time. Resolve
+	// it to an IDP-specific stable identifier so that we can support subject renames.
+	identity := &auth.Identity{
+		Provider:  nil,
+		UserID:    subject,
+		HumanName: subject,
+	}
+	if flags.Bool(ctx, s.featureFlags, flags.IDPResolver) {
+		identity, err = s.idClient.Resolve(ctx, subject)
+		if err != nil {
+			zerolog.Ctx(ctx).Error().Err(err).Msg("error resolving identity")
+			return nil, util.UserVisibleError(codes.NotFound, "could not find identity %q", subject)
+		}
+	}
+
+	// Verify if user exists
+	if _, err := s.store.GetUserBySubject(ctx, identity.UserID); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, util.UserVisibleError(codes.NotFound, "User not found")
+		}
+		return nil, status.Errorf(codes.Internal, "error getting user: %v", err)
+	}
+
+	if err := s.authzClient.Delete(ctx, identity.String(), role, targetProject); err != nil {
+		return nil, status.Errorf(codes.Internal, "error writing role assignment: %v", err)
+	}
+	prj := targetProject.String()
+	return &minder.RemoveRoleResponse{
+		RoleAssignment: &minder.RoleAssignment{
+			Role:    role.String(),
+			Subject: identity.Human(),
+			Project: &prj,
+		},
+	}, nil
+}
+
+// isEmail checks if the subject is an email address or not
+func isEmail(subject string) bool {
+	// Define the regular expression for validating an email address
+	const emailRegexPattern = `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
+	emailRegex := regexp.MustCompile(emailRegexPattern)
```

Comment:
Compile the email regexp just once to reuse it across calls

```suggestion

var emailRegexPattern *regexp.Regexp
// isEmail checks if the subject is an email address or not
func isEmail(subject string) bool {
	// Define the regular expression for validating an email address
	const emailRegexPattern = `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
	if emailRegexPattern == nil {
	  emailRegex = regexp.MustCompile(emailRegexPattern)
	}
```

Reply:
Good tip! Let me update this manually though as I think some of the env names in the suggestion are not correct.

Reply:
@puerco the email regex was actually removed to let the e-mail sending service verify it, not minder (it's unlikely we'd ever get the regex right for all weird cases)